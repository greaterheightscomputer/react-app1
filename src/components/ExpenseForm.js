import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// const now = moment();
// console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
    constructor(props){ //constructor(props) its will allow ExpenseForm to access the super class (EditExpensePage) props if its exist else its return null  
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '', 
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }    
    onDescriptionChange = (e) => {
        const description = e.target.value;
        // console.log(description)
        this.setState(() => ({ description }));
    };
    onNoteChange = (e) => {
        // const note = e.target.value;
        e.persist();
        this.setState(() => ({ note: e.target.value }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){  //!amount its will allow user to delete the text field of amount 
            this.setState(() => ({ amount }));
        }
    };
    onDateChange = (createdAt) => { //api of onDateChange get called with date as an argument since createdAt represent date that is why it is pass as an argument
        if (createdAt) {  //if statement will prevent the user from delete the date field
            this.setState(() => ({ createdAt }));
        };        
    };
    onFocusChange = ({ focused }) => { //destructuring of {focused}
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault();  //prevent full page refresh
        if (!this.state.description || !this.state.amount){
            this.setState(() => ({ error: 'Please provide description or amount' }));            
        }else{
            this.setState(() => ({ error: '' }));  
            this.props.onSubmits({  //we pass the data out of the ExpenseForm by calling the props from the parent component we did that beco's we want to use the ExpenseForm on AddExpensePage and later on EditExpensePage 
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),     //valueOf() return moment in milliseconds value
                note: this.state.note
            });
        }
    };
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text" 
                        placeholder="Description" 
                        autoFocus  
                        value={this.state.description} 
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type="text" 
                        placeholder="Amount" 
                        value={this.state.amount} 
                        onChange={this.onAmountChange} 
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt}     //display the current date as the app render on the browser since the initial value of createdAt is set to moment()
                        onDateChange={this.onDateChange}   //it is use to change the application state when the user select new date 
                        focused={this.state.calendarFocused} //focused by default is false beco's user is not interreacting with the datepicker
                        onFocusChange={this.onFocusChange} //onFocusChange() means user is now changing the app state by intereacting with calenderFocus property by changing from false to true
                        numberOfMonths={1}    //it will render only one month
                        isOutsideRange={() => false}    //it will allow user to select or pick date back in time
                    />
                    <textarea 
                        placeholder="Add a note for your expense (optional)" 
                        value={this.state.note} 
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}