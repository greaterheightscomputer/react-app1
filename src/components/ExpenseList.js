import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

//The Component that Fetch data from database
// React component
export const ExpenseList = (props) => (
    <div className="content-container">    
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>  
        <div className="list-body">       
            {
                props.expenses.length === 0 ? (   //expenses is from mapStateToProps 
                    <div className="list-item list-item--error">
                        <span>No expenses</span>
                    </div>
                ) : (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense}/>                    
                    })
                )
            }
        </div>
    </div>
);

//Redux store
const mapStateToProps = (state) => { //mapStateToProps it is a function that fetch data from the redux store
    return {
        expenses: selectExpenses(state.expenses, state.filters)
        // expenses: state.expenses,
        // filters: state.filters
    };
};

// Higher Order Component
// connect -> connect the component with redux store
// this connection allow our component to access any data from the redux store
export default connect(mapStateToProps)(ExpenseList); //combinating the store with the component

// const ConnectedExpensesList = connect((state) => {
//     return {
//         expenses: state.expenses
//     };
// })(ExpenseList);
// export default ConnectedExpensesList;