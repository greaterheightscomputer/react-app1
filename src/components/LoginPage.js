import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => ( //destructure startLogin
//export const LoginPage = (props) => (
// <button onClick={props.startLogin}>Login</button>
    <div>
        <button onClick={startLogin}>Login</button>        
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage); 