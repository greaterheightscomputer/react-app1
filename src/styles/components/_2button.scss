.button { //base class
    background: $blue;
    border: none;
    color: white;
    display: inline-block;
    font-size: $font-size-large;
    font-weight: 300;
    line-height: 1;
    padding: $s-size;
    text-decoration: none;
}

.button--link { //modifier class
    background: none;
}

.button--secondary {
    background: #888888;
}