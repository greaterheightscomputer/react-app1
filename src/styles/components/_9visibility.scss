.show-for-mobile {
    @media (min-width: $desktop-breakpoint) {
        display: none;        
    }
}

.show-for-desktop {
    @media (max-width: $desktop-breakpoint - .01rem) { //.01rem will make sure that the two selector does not occur at the sametime 
        display: none;        
    }
}