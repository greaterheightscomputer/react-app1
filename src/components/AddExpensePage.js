import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmited = (expense) => {           //event method      
           this.props.startAddExpense(expense)   //addExpense is an Action Generator Function will pass data to reducer then reducer will pass the data to redux store
           this.props.history.push('/');    //history.push('/') its will take us back to the Dashboard after submittion 
    };
    render() {
        return(
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm       //the <ExpenseForm/> component will render form to the browser 
                    onSubmit={this.onSubmited} //the onSubmit props will get data from ExpenseForm component on this.props.onSubmit() function and pass in the values set properties to addExpense Action Generator Function and go back to the Dashboard
                    />
                </div>                
            </div>
        );
    }
} 
// const AddExpensePage = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm
//            onSubmit={(expense) => {
//             console.log(expense)
//             props.dispatch(addExpense(expense));  //on this line we are doing two things, we calling props.dispatch() which can easily be a spy or fake function but addExpense(expense) cannot be a spy or fake function beco's its imported and it is not pass in as a props, as a result will need to use a function that is pass to connect() which is called mapDispatchToProps which will help us to pass addExpense(expense) to props for easy testing.           
//             props.history.push('/');
//            }}        
//         />
//     </div>
// );

// mapDispatchToProps is a way to return your dispatch function, allowing you to extract or sperate them from the component itself. It will also allow you to pass Action Generator Function to props for easy testing.   
const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);