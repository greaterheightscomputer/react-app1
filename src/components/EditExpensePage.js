import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmited = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');   
    };
    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/'); 
    };
    render(){
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm   
                        expense={this.props.expense} //its will get expense properties from mapStateToProps property expense (redux store) and set on the application state as a result once user click on the description property link on the DashboardPage its will populate the text fields with values.  
                        onSubmit={this.onSubmited} //its get the edited data from the sub-class ExpenseForm and pass to editEpense() Action Generator Function
                    />                
                    <button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
                </div>                
            </div>
        );
    }
};

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id) //if find() return true its means its has find a match then its store that particular expense properities in expense object          
});

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))    
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

