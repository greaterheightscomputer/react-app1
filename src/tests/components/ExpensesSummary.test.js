import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should correctly render ExpensesSummary with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={245}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary with multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={24} expensesTotal={45678900}/>);
    expect(wrapper).toMatchSnapshot();
});