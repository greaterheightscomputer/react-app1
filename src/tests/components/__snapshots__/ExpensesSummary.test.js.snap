// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  <h1>
     Viewing 
    1
     
    expense
     totalling 
    ₦2.34
  </h1>
</div>
`;

exports[`should correctly render ExpensesSummary with multiple expense 1`] = `
<div>
  <h1>
     Viewing 
    21
     
    expenses
     totalling 
    ₦2,233.34
  </h1>
</div>
`;
